src/ec/system76/ec/system76_ec.c

```diff
@@ -148,3 +187,156 @@ void system76_ec_set_bat_threshold(enum bat_threshold_type type, uint8_t value)
+
+	return ret_i;
+}
+
+struct smfi_spi_cmd_hdr {
+	uint8_t flags;
+	uint8_t len;
+} __packed;
+
+static uint32_t flash_read(uint8_t *dest, uint32_t len) {
+	struct smfi_spi_cmd_hdr read_cmd = {
+		CMD_SPI_FLAG_READ,
+		SYSTEM76_EC_SIZE - 4,
+	};
+
+	uint32_t addr = 0;
+	uint32_t bytes_read = 0;
+	int i;
+
+	while (addr < len) {
+		wait_us(10000, system76_ec_read(REG_CMD) == CMD_FINISHED);
+
+		if (system76_ec_smfi_cmd(CMD_SPI, sizeof(read_cmd) / sizeof(uint8_t), (uint8_t *)&read_cmd)) {
+			printk(BIOS_DEBUG, "EC flash read failed!\n");
+			return addr + 1;
+		}
+
+		/* Check how many data bytes were actually read */
+		bytes_read = system76_ec_read(REG_DATA + 1);
+
+		/* Sanity check */
+		if (bytes_read == 0 || bytes_read > SYSTEM76_EC_SIZE - 4)
+			return 0;
+
+		/* Read data bytes, index should be valid due to length test above */
+		for (i = 0; i < bytes_read; i++)
+			dest[addr + i] = system76_ec_read(REG_DATA + 2 + i);
+
+		addr += bytes_read;
+
+		/* Check if this was the last chunk */
+		if (bytes_read < SYSTEM76_EC_SIZE - 4)
+			break;
+	}
+
+	return addr + 1;
+}
+
+static void system76_ec_fw_sync(void *unused)
+{
+	size_t image_sz;
+	char *image;
+	char img_board_str[64];
+	char img_version_str[64];
+	char cur_board_str[64];
+	char cur_version_str[64];
+
+	/* EC update is loaded from verified CBFS */
+	image = cbfs_map("ec.rom", &image_sz);
+
+	if (image == NULL) {
+		printk(BIOS_DEBUG, "No EC firmware update found.\n");
+		return;
+	}
+
+	printk(BIOS_DEBUG, "EC update found (%ld bytes)\n", image_sz);
+
+	/* Sanity checks */
+	if (!firmware_str(image, image_sz, "76EC_BOARD=", img_board_str, ARRAY_SIZE(img_board_str))) {
+		printk(BIOS_DEBUG, "Could not determine EC update target board.\n");
+		return;
+	}
+
+	if (!firmware_str(image, image_sz, "76EC_VERSION=", img_version_str, ARRAY_SIZE(img_version_str))) {
+		printk(BIOS_DEBUG, "Could not determine EC update version.\n");
+		return;
+	}
+
+	printk(BIOS_DEBUG, "EC update target: %s\n", img_board_str);
+	printk(BIOS_DEBUG, "EC update version: %s\n", img_version_str);
+
+	system76_ec_read_board((uint8_t *)cur_board_str);
+	system76_ec_read_version((uint8_t *)cur_version_str);
+
+	if (strcmp(img_board_str, cur_board_str)) {
+		printk(BIOS_DEBUG, "EC update mismatch detected! Found %s, expected %s\n",
+			img_board_str, cur_board_str);
+		return;
+	}
+
+	printk(BIOS_DEBUG, "EC current version: %s\n", cur_version_str);
+	if (!strcmp(img_version_str, cur_version_str)) {
+		printk(BIOS_DEBUG, "EC update not needed.\n");
+		return;
+	} else {
+		printk(BIOS_DEBUG, "EC update required!\n");
+	}
+
+	/* Jump to Scratch ROM */
+	struct smfi_spi_cmd_hdr smfi_cmd = {
+		CMD_SPI_FLAG_SCRATCH,
+		0,
+	};
+
+	/* If we failed to jump to scratch ROM, then we can probably continue booting. */
+	if (system76_ec_smfi_cmd(CMD_SPI, sizeof(smfi_cmd) / sizeof(uint8_t), (uint8_t *)&smfi_cmd)) {
+		printk(BIOS_ERR, "EC update failed!\n");
+		return;
+	}
+
+	uint8_t current_rom[SPI_FLASH_SIZE] = { 0xFF };
```

Comment:
Don't do it like that. You basically request 128K of memory to be allocated on stack. coreboot's stake is only a couple of kilobytes...

Do a malloc instead and free the buffer after use. Then the buffer will be used as heap.

@mkopec

Reply:
makes sense, applied your suggestion in https://github.com/Dasharo/coreboot/pull/456/commits/e34fee6c8657ee449fdf5824651ad457f7c81f01 but it fails as well with this error:

```
memalign(boundary=8, size=131072): failed: Tried to round up free_mem_ptr 0x76b0ba30 to 0x76b2ba30
but free_mem_end_ptr is 0x76b1ba30
```

I suppose we don't actually need to allocate the full 128KB, for verification after update we can just compare chunk by chunk

Reply:
As you can see, coreboot heap is 64KiB too short to allocate full flash size.

> I suppose we don't actually need to allocate the full 128KB, for verification after update we can just compare chunk by chunk

Probably better idea.